lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs

```diff
@@ -621,4 +627,8 @@ postAnyAddress net addrData = do
                       Left err -> (True, TextDecodingError $ prettyErrValidateScript err)
                       Right _ -> (False, TextDecodingError "")
               _ -> (False, TextDecodingError "")
-      (Right discriminant) = CA.mkNetworkDiscriminant netTag
+      discriminant = case CA.mkNetworkDiscriminant netTag of
```

Comment:
I don't know… 🤔 I don't think that catching the `Left` case with `PatternMatchFail` adds more information than, well, just having the pattern match fail. I believe that `HasCallStack` is a more economical way to gain information in the case of failure.

In any case, I would argue that using `let (Right r) = exp` is fine *if* the proof obligation that `exp` is always a `Right` has been taken care of somewhere else. Here is a specific, related example with QuickCheck and NonEmpty lists that I recently ran into:

```hs
import Test.QuickCheck (Gen, choose)
import qualified Data.List.NonEmpty as NE

ys :: NonEmpty X

foo :: Gen (NonEmpty X)
foo = do
  m  <- choose (1,max 1 (NE.length ys -3))
  xs <- NE.take m ys
  return $ NE.fromList xs
```
Here, `NE.fromList` is a partial function, but the generator is made such that the proof obligation `m >= 1` holds. My point is that nothing would be gained by using `NE.nonEmpty :: [a] -> Maybe (NonEmpty a)` and pattern matching the `Nothing` case — this doesn't make the function any more total, as we still would have to dismiss the `Nothing` case on the grounds that it can never occur by virtue of our proof obligation (here `m>=1`). Put differently, handling the `Nothing` case adds more code, but doesn't add any value, as the proof obligation does not go away.

I would apply a similar line of reasoning to the proposed warnings. 🤓 Happy to chat about it, though!

Reply:
> I don't know… thinking I don't think that catching the Left case with `PatternMatchFail` adds more information than, well, just having the pattern match fail. 

I agree. It wasn't my primary goal to add more information to the failure message. The purpose of this PR is to make **incomplete pattern matches** a **compiler warning** by introducing the `-Wincomplete-uni-patterns` flag. This specific change is a way to achieve the same behavior as before without a compiler warning.

> In any case, I would argue that using `let (Right r) = exp` is fine if the proof obligation that exp is always a `Right` has been taken care of somewhere else. 

I agree that we don't always push type system far enough to express certain invariants (example: [Ghosts of Departed Proofs (Functional Pearl)](https://kataskeue.com/gdp.pdf)). Instead of trying to express as much as possible in the types (whatever complexity it takes) we are trying to find a proper balance between expressiveness and code simplicity. With such mindset partial functions are unavoidable and are useful occasionally.

The point of this PR is not to argue "to use partial functions (incomplete pattern matches) or not" but to flip the default:
while its still possible to use partiality, let compiler issue a warning, to remind developer about their obligation to prove the safety by:

- Leaving a comment explaining why partiality is actually safe in this specific case;
- Using explicitly unsafe functions like https://hackage.haskell.org/package/relude-1.0.0.1/docs/Relude-Unsafe.html 

Instead of:

```haskell
let (Right discriminant) = CA.mkNetworkDiscriminant netTag -- Compiler warning
```

I encourage to:

```haskell
-- Network discriminant is always present because
-- of runtime check in the decoder
let discriminant = Unsafe.fromRight $ CA.mkNetworkDiscriminant netTag
```
or even 

```haskell
let discriminant = Unsafe.fromRight
        "Network discriminant is always present because \
        \of runtime check in the decoder" $
        CA.mkNetworkDiscriminant netTag
```

Such discipline leads to a better error messages in "impossible cases" and helps readers to understand the motivation and assumptions associated with the usage of unsafe partial code.

This PR is a first step in this direction.

Reply:
Just as an aside, in the above example we can avoid the partial function altogether: (Since `NE.head` is total and `take n | n <= 0` evaluates to `[]`)

```patch
  import Test.QuickCheck (Gen, choose)
  import qualified Data.List.NonEmpty as NE

  ys :: NonEmpty X

  foo :: Gen (NonEmpty X)
  foo = do
-     m <- choose (1,max 1 (NE.length ys -3))
-     xs <- NE.take m ys
-     return $ NE.fromList xs
+     m <- choose (0, NE.length ys - 4)
+     return $ NE.head ys :| take m NE.tail ys
```